modules/dag-shared/src/main/scala/org/tessellation/dag/validators/TransactionValidator.scala

```
@@ -0,0 +1,208 @@
+    } yield propertiesValidation.product(balancesValidation).map(_ => signedTransactions)
+
+  private def validateTransactionProperties(
+    signedTransaction: Signed[Transaction]
+  ): F[ValidationResult[Signed[Transaction]]] =
+    for {
+      staticValidation <- Async[F].delay(
+        validateAddressInequality(signedTransaction.value)
+      )
+      signatureValidation <- validateSourceSignature(signedTransaction)
+      sourceValidation <- validateSourceAddress(signedTransaction)
+      lastTxRefValidation <- validateLastTransactionRef(signedTransaction.value)
+    } yield
+      staticValidation
+        .product(signatureValidation)
+        .product(sourceValidation)
+        .product(lastTxRefValidation)
+        .map(_ => signedTransaction)
+
+  private def validateLastTransactionRef(tx: Transaction): F[ValidationResult[Transaction]] =
+    getLastAcceptedTransactionRef(tx.source).map { lastTxRef =>
+      lazy val comparedOrdinals = tx.parent.ordinal.coerce.compare(lastTxRef.ordinal.coerce)
+      lazy val nonZeroOrdinalHasEmptyHash = tx.parent.ordinal.coerce > 0L && tx.parent.hash.coerce.isEmpty
+      lazy val ordinalIsLower = comparedOrdinals < 0L
+      lazy val hashIsNotEqual = comparedOrdinals == 0L && tx.parent != lastTxRef
+
+      if (nonZeroOrdinalHasEmptyHash)
+        NonZeroOrdinalButEmptyHash(tx).invalidNel
+      else if (ordinalIsLower)
+        ParentTxRefOrdinalLowerThenStoredLastTxRef(tx).invalidNel
+      else if (hashIsNotEqual)
+        SameOrdinalButDifferentHashForLastTxRef(tx).invalidNel
+      else
+        tx.validNel
+    }
+
+  private def validateSourceAddress(
+    signedTransaction: Signed[Transaction]
+  ): F[ValidationResult[Transaction]] =
+    signedTransaction.proofs.traverse {
+      case SignatureProof(id, _) => {
+        for {
+          publicKey <- id.hex.toPublicKey
+          address = publicKey.toAddress
+        } yield address == signedTransaction.value.source
+      }.handleError(_ => false)
```

This will swallow the error. Is that supposed?

---

Yep, the error returned in the following lines is descriptive enough.

---

I mean we could log it, but then we would need a logger in the validator, maybe the failing validation case is good enough?

---

but here you have an error that is different than inequality (something throws under the hood) and this error is gone (i.e. kryo issue)

---

Yeah I know. I'm just wondering though if we should log everything everywhere 🤔.